File: src/algorithms/dynamic-programming/lis-memoization/sources/lis-memoization.py
# LIS (Longest Increasing Subsequence) memoization — top-down recursion with cached subproblems

def lis_memoization(sequence: list[int]) -> int:
    sequence_length = len(sequence)  # @step:initialize
    if sequence_length == 0:  # @step:initialize
        return 0  # @step:initialize
    memo: dict[int, int] = {}  # @step:initialize

    def lis(start_index: int) -> int:
        if start_index in memo:  # @step:read-cache
            return memo[start_index]  # @step:read-cache
        # @step:push-call
        max_length = 1  # @step:compute-cell
        for next_index in range(start_index + 1, sequence_length):  # @step:compute-cell
            if sequence[next_index] > sequence[start_index]:  # @step:compute-cell
                sub_length = 1 + lis(next_index)  # @step:compute-cell
                if sub_length > max_length:  # @step:compute-cell
                    max_length = sub_length  # @step:compute-cell
        memo[start_index] = max_length  # @step:compute-cell
        return max_length  # @step:pop-call

    result = 0  # @step:compute-cell
    for start_index in range(sequence_length):  # @step:compute-cell
        lis_length = lis(start_index)  # @step:compute-cell
        if lis_length > result:  # @step:compute-cell
            result = lis_length  # @step:compute-cell

    return result  # @step:complete

File: src/algorithms/dynamic-programming/subsequence/lis-memoization/sources/LisMemoization.cpp
// LIS (Longest Increasing Subsequence) memoization — top-down recursion with cached subproblems

#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>

int lis(const std::vector<int>& sequence, int startIndex, std::unordered_map<int, int>& memo) {
    auto it = memo.find(startIndex);
    if (it != memo.end()) return it->second; // @step:read-cache
    // @step:push-call
    int maxLength = 1; // @step:compute-cell
    int sequenceLength = sequence.size();
    for (int nextIndex = startIndex + 1; nextIndex < sequenceLength; nextIndex++) {
        // @step:compute-cell
        if (sequence[nextIndex] > sequence[startIndex]) {
            // @step:compute-cell
            int subLength = 1 + lis(sequence, nextIndex, memo); // @step:compute-cell
            if (subLength > maxLength) {
                // @step:compute-cell
                maxLength = subLength; // @step:compute-cell
            }
        }
    }
    memo[startIndex] = maxLength; // @step:compute-cell
    return maxLength; // @step:pop-call
}

int lisMemoization(const std::vector<int>& sequence) {
    // @step:initialize
    int sequenceLength = sequence.size(); // @step:initialize
    if (sequenceLength == 0) return 0; // @step:initialize
    std::unordered_map<int, int> memo; // @step:initialize
    int result = 0; // @step:compute-cell
    for (int startIndex = 0; startIndex < sequenceLength; startIndex++) {
        // @step:compute-cell
        int lisLength = lis(sequence, startIndex, memo); // @step:compute-cell
        if (lisLength > result) {
            // @step:compute-cell
            result = lisLength; // @step:compute-cell
        }
    }
    return result; // @step:complete
}

#ifndef TESTING
int main() {
    std::vector<int> sequence = {10, 9, 2, 5, 3, 7, 101, 18};
    int result = lisMemoization(sequence);
    std::cout << "LIS length: " << result << std::endl;
    return 0;
}
#endif
